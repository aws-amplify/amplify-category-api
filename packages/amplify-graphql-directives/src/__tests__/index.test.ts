import { parse } from 'graphql';
import * as Directives from '..';
import type { Directive } from '..';
import { AppSyncDirectives, DefaultDirectives, V1Directives, V2Directives } from '..';

describe('Directive Definitions', () => {
  // remove lists of Directives, i.e. AppSyncDirectives, DefaultDirectives, V1Directives, and V2Directives
  test.each(Object.entries(Directives).filter(([_, directive]) => !Array.isArray(directive)) as unknown as [string, Directive][])(
    '%s',
    (_, directive: Directive) => {
      // assert valid graphql syntax
      expect(() => parse(directive.definition)).not.toThrow();

      // assert no changes to directive
      expect(directive).toMatchSnapshot();
    },
  );

  test('include correct directives in AppSync directives', () => {
    expect(AppSyncDirectives.map((directive) => directive.name)).toMatchSnapshot();
  });

  test('no negative interactions for AppSync directives', () => {
    const directives = AppSyncDirectives.map((directive) => directive.definition).join('\n');

    // asserts directives can be parsed together
    expect(() => parse(directives)).not.toThrow();
  });

  test('include correct directives in V2 directives', () => {
    expect(V2Directives.map((directive) => directive.name)).toMatchSnapshot();
  });

  test('no negative interactions for V2 directives', () => {
    const directives = V2Directives.map((directive) => directive.definition).join('\n');

    // asserts directives can be parsed together
    expect(() => parse(directives)).not.toThrow();
  });

  test('include correct directives in default directives', () => {
    expect(DefaultDirectives.map((directive) => directive.name)).toMatchSnapshot();
  });

  test('no negative interactions for default directives', () => {
    const directives = DefaultDirectives.map((directive) => directive.definition).join('\n');

    // asserts directives can be parsed together
    expect(() => parse(directives)).not.toThrow();
  });

  test('include correct directives in V1 directives', () => {
    expect(V1Directives.map((directive) => directive.name)).toMatchSnapshot();
  });

  test('no negative interactions for V1 directives', () => {
    const directives = V1Directives.map((directive) => directive.definition).join('\n');

    // asserts directives can be parsed together
    expect(() => parse(directives)).not.toThrow();
  });
});
