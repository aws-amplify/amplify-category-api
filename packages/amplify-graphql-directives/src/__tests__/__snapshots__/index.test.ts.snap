// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Directive Definitions AuthDirective 1`] = `
Object {
  "defaults": Object {},
  "definition": "
  directive @auth(rules: [AuthRule!]!) on OBJECT | FIELD_DEFINITION
  input AuthRule {
    allow: AuthStrategy!
    provider: AuthProvider
    identityClaim: String
    groupClaim: String
    ownerField: String
    groupsField: String
    groups: [String]
    operations: [ModelOperation]
  }
  enum AuthStrategy {
    owner
    groups
    private
    public
    custom
  }
  enum AuthProvider {
    apiKey
    iam
    identityPool
    oidc
    userPools
    function
  }
  enum ModelOperation {
    create
    update
    delete
    read
    list
    get
    sync
    listen
    search
  }
",
  "name": "auth",
}
`;

exports[`Directive Definitions AuthDirectiveV1 1`] = `
Object {
  "defaults": Object {},
  "definition": "
  directive @auth(rules: [AuthRule!]!) on OBJECT | FIELD_DEFINITION
  input AuthRule {
    # Specifies the auth rule's strategy. Allowed values are 'owner', 'groups', 'public', 'private'.
    allow: AuthStrategy!

    # Legacy name for identityClaim
    identityField: String @deprecated(reason: \\"The 'identityField' argument is replaced by the 'identityClaim'.\\")

    # Specifies the name of the provider to use for the rule. This overrides the default provider
    # when 'public' and 'private' AuthStrategy is used. Specifying a provider for 'owner' or 'groups'
    # are not allowed.
    provider: AuthProvider

    # Specifies the name of the claim to look for on the request's JWT token
    # from Cognito User Pools (and in the future OIDC) that contains the identity
    # of the user. If 'allow' is 'groups', this value should point to a list of groups
    # in the claims. If 'allow' is 'owner', this value should point to the logged in user identity string.
    # Defaults to \\"cognito:username\\" for Cognito User Pools auth.
    identityClaim: String

    # Allows for custom config of 'groups' which is validated against the JWT
    # Specifies a static list of groups that should have access to the object
    groupClaim: String

    # Allowed when the 'allow' argument is 'owner'.
    # Specifies the field of type String or [String] that contains owner(s) that can access the object.
    ownerField: String # defaults to \\"owner\\"
    # Allowed when the 'allow' argument is 'groups'.
    # Specifies the field of type String or [String] that contains group(s) that can access the object.
    groupsField: String

    # Allowed when the 'allow' argument is 'groups'.
    # Specifies a static list of groups that should have access to the object.
    groups: [String]

    # Specifies operations to which this auth rule should be applied.
    operations: [ModelOperation]

    # Deprecated. It is recommended to use the 'operations' arguments.
    queries: [ModelQuery] @deprecated(reason: \\"The 'queries' argument will be replaced by the 'operations' argument in a future release.\\")

    # Deprecated. It is recommended to use the 'operations' arguments.
    mutations: [ModelMutation]
      @deprecated(reason: \\"The 'mutations' argument will be replaced by the 'operations' argument in a future release.\\")
  }
  enum AuthStrategy {
    owner
    groups
    private
    public
  }
  enum AuthProvider {
    apiKey
    iam
    oidc
    userPools
  }
  enum ModelOperation {
    create
    update
    delete
    read
  }
  enum ModelQuery @deprecated(reason: \\"ModelQuery will be replaced by the 'ModelOperation' in a future release.\\") {
    get
    list
  }
  enum ModelMutation @deprecated(reason: \\"ModelMutation will be replaced by the 'ModelOperation' in a future release.\\") {
    create
    update
    delete
  }
",
  "name": "auth",
}
`;

exports[`Directive Definitions AwsApiKeyDirective 1`] = `
Object {
  "defaults": Object {},
  "definition": "
  directive @aws_api_key on FIELD_DEFINITION | OBJECT
",
  "name": "aws_api_key",
}
`;

exports[`Directive Definitions AwsAuthDirective 1`] = `
Object {
  "defaults": Object {},
  "definition": "
  directive @aws_auth(cognito_groups: [String!]!) on FIELD_DEFINITION
",
  "name": "aws_auth",
}
`;

exports[`Directive Definitions AwsCognitoUserPoolsDirective 1`] = `
Object {
  "defaults": Object {},
  "definition": "
  directive @aws_cognito_user_pools(cognito_groups: [String!]) on FIELD_DEFINITION | OBJECT
",
  "name": "aws_cognito_user_pools",
}
`;

exports[`Directive Definitions AwsIamDirective 1`] = `
Object {
  "defaults": Object {},
  "definition": "
  directive @aws_iam on FIELD_DEFINITION | OBJECT
",
  "name": "aws_iam",
}
`;

exports[`Directive Definitions AwsLambdaDirective 1`] = `
Object {
  "defaults": Object {},
  "definition": "
  directive @aws_lambda on FIELD_DEFINITION | OBJECT
",
  "name": "aws_lambda",
}
`;

exports[`Directive Definitions AwsOidcDirective 1`] = `
Object {
  "defaults": Object {},
  "definition": "
  directive @aws_oidc on FIELD_DEFINITION | OBJECT
",
  "name": "aws_oidc",
}
`;

exports[`Directive Definitions AwsSubscribeDirective 1`] = `
Object {
  "defaults": Object {},
  "definition": "
  directive @aws_subscribe(mutations: [String!]!) on FIELD_DEFINITION
",
  "name": "aws_subscribe",
}
`;

exports[`Directive Definitions BelongsToDirective 1`] = `
Object {
  "defaults": Object {},
  "definition": "
  directive @belongsTo(fields: [String!], references: [String!]) on FIELD_DEFINITION
",
  "name": "belongsTo",
}
`;

exports[`Directive Definitions ConnectionDirectiveV1 1`] = `
Object {
  "defaults": Object {},
  "definition": "
  directive @connection(
    name: String
    keyField: String
    sortField: String
    keyName: String
    limit: Int
    fields: [String!]
  ) on FIELD_DEFINITION
",
  "name": "connection",
}
`;

exports[`Directive Definitions DefaultDirective 1`] = `
Object {
  "defaults": Object {},
  "definition": "
  directive @default(value: String!) on FIELD_DEFINITION
",
  "name": "default",
}
`;

exports[`Directive Definitions DeprecatedDirective 1`] = `
Object {
  "defaults": Object {},
  "definition": "
  directive @deprecated(reason: String) on FIELD_DEFINITION | INPUT_FIELD_DEFINITION | ENUM | ENUM_VALUE
",
  "name": "deprecated",
}
`;

exports[`Directive Definitions FunctionDirective 1`] = `
Object {
  "defaults": Object {},
  "definition": "
  directive @function(name: String!, region: String, accountId: String) repeatable on FIELD_DEFINITION
",
  "name": "function",
}
`;

exports[`Directive Definitions FunctionDirectiveV1 1`] = `
Object {
  "defaults": Object {},
  "definition": "
  directive @function(name: String!, region: String) repeatable on FIELD_DEFINITION
",
  "name": "function",
}
`;

exports[`Directive Definitions HasManyDirective 1`] = `
Object {
  "defaults": Object {
    "limit": 100,
  },
  "definition": "
  directive @hasMany(indexName: String, fields: [String!], references: [String!], limit: Int = 100) on FIELD_DEFINITION
",
  "name": "hasMany",
}
`;

exports[`Directive Definitions HasOneDirective 1`] = `
Object {
  "defaults": Object {},
  "definition": "
  directive @hasOne(fields: [String!], references: [String!]) on FIELD_DEFINITION
",
  "name": "hasOne",
}
`;

exports[`Directive Definitions HttpDirective 1`] = `
Object {
  "defaults": Object {
    "headers": Array [],
    "method": "GET",
  },
  "definition": "
  directive @http(method: HttpMethod = GET, url: String!, headers: [HttpHeader] = []) on FIELD_DEFINITION
  enum HttpMethod {
    GET
    POST
    PUT
    DELETE
    PATCH
  }
  input HttpHeader {
    key: String
    value: String
  }
",
  "name": "http",
}
`;

exports[`Directive Definitions HttpDirectiveV1 1`] = `
Object {
  "defaults": Object {
    "headers": Array [],
    "method": "GET",
  },
  "definition": "
  directive @http(method: HttpMethod = GET, url: String!, headers: [HttpHeader] = []) on FIELD_DEFINITION
  enum HttpMethod {
    GET
    POST
    PUT
    DELETE
    PATCH
  }
  input HttpHeader {
    key: String
    value: String
  }
",
  "name": "http",
}
`;

exports[`Directive Definitions IndexDirective 1`] = `
Object {
  "defaults": Object {},
  "definition": "
  directive @index(name: String, sortKeyFields: [String], queryField: String) repeatable on FIELD_DEFINITION
",
  "name": "index",
}
`;

exports[`Directive Definitions KeyDirectiveV1 1`] = `
Object {
  "defaults": Object {},
  "definition": "
  directive @key(name: String, fields: [String!]!, queryField: String) repeatable on OBJECT
",
  "name": "key",
}
`;

exports[`Directive Definitions ManyToManyDirective 1`] = `
Object {
  "defaults": Object {
    "limit": 100,
  },
  "definition": "
  directive @manyToMany(relationName: String!, limit: Int = 100) on FIELD_DEFINITION
",
  "name": "manyToMany",
}
`;

exports[`Directive Definitions MapsToDirective 1`] = `
Object {
  "defaults": Object {},
  "definition": "
  directive @mapsTo(name: String!) on OBJECT
",
  "name": "mapsTo",
}
`;

exports[`Directive Definitions ModelDirective 1`] = `
Object {
  "defaults": Object {},
  "definition": "
  directive @model(
    queries: ModelQueryMap
    mutations: ModelMutationMap
    subscriptions: ModelSubscriptionMap
    timestamps: TimestampConfiguration
  ) on OBJECT
  input ModelMutationMap {
    create: String
    update: String
    delete: String
  }
  input ModelQueryMap {
    get: String
    list: String
  }
  input ModelSubscriptionMap {
    onCreate: [String]
    onUpdate: [String]
    onDelete: [String]
    level: ModelSubscriptionLevel
  }
  enum ModelSubscriptionLevel {
    off
    public
    on
  }
  input TimestampConfiguration {
    createdAt: String
    updatedAt: String
  }
",
  "name": "model",
}
`;

exports[`Directive Definitions ModelDirectiveV1 1`] = `
Object {
  "defaults": Object {},
  "definition": "
  directive @model(
    queries: ModelQueryMap
    mutations: ModelMutationMap
    subscriptions: ModelSubscriptionMap
    timestamps: TimestampConfiguration
  ) on OBJECT
  input ModelMutationMap {
    create: String
    update: String
    delete: String
  }
  input ModelQueryMap {
    get: String
    list: String
  }
  input ModelSubscriptionMap {
    onCreate: [String]
    onUpdate: [String]
    onDelete: [String]
    level: ModelSubscriptionLevel
  }
  enum ModelSubscriptionLevel {
    off
    public
    on
  }
  input TimestampConfiguration {
    createdAt: String
    updatedAt: String
  }
",
  "name": "model",
}
`;

exports[`Directive Definitions PredictionsDirective 1`] = `
Object {
  "defaults": Object {},
  "definition": "
  directive @predictions(actions: [PredictionsActions!]!) on FIELD_DEFINITION
  enum PredictionsActions {
    identifyText
    identifyLabels
    convertTextToSpeech
    translateText
  }
",
  "name": "predictions",
}
`;

exports[`Directive Definitions PredictionsDirectiveV1 1`] = `
Object {
  "defaults": Object {},
  "definition": "
  directive @predictions(actions: [PredictionsActions!]!) on FIELD_DEFINITION
  enum PredictionsActions {
    identifyText
    identifyLabels
    convertTextToSpeech
    translateText
  }
",
  "name": "predictions",
}
`;

exports[`Directive Definitions PrimaryKeyDirective 1`] = `
Object {
  "defaults": Object {},
  "definition": "
  directive @primaryKey(sortKeyFields: [String]) on FIELD_DEFINITION
",
  "name": "primaryKey",
}
`;

exports[`Directive Definitions RefersToDirective 1`] = `
Object {
  "defaults": Object {},
  "definition": "
  directive @refersTo(name: String!) on OBJECT | FIELD_DEFINITION
",
  "name": "refersTo",
}
`;

exports[`Directive Definitions SearchableDirective 1`] = `
Object {
  "defaults": Object {},
  "definition": "
  directive @searchable(queries: SearchableQueryMap) on OBJECT
  input SearchableQueryMap {
    search: String
  }
",
  "name": "searchable",
}
`;

exports[`Directive Definitions SearchableDirectiveV1 1`] = `
Object {
  "defaults": Object {},
  "definition": "
  directive @searchable(queries: SearchableQueryMap) on OBJECT
  input SearchableQueryMap {
    search: String
  }
",
  "name": "searchable",
}
`;

exports[`Directive Definitions SqlDirective 1`] = `
Object {
  "defaults": Object {},
  "definition": "
  directive @sql(statement: String, reference: String) on FIELD_DEFINITION
",
  "name": "sql",
}
`;

exports[`Directive Definitions VersionedDirectiveV1 1`] = `
Object {
  "defaults": Object {
    "versionField": "version",
    "versionInput": "expectedVersion",
  },
  "definition": "
  directive @versioned(versionField: String = \\"version\\", versionInput: String = \\"expectedVersion\\") on OBJECT
",
  "name": "versioned",
}
`;

exports[`Directive Definitions include correct directives in AppSync directives 1`] = `
Array [
  "aws_api_key",
  "aws_auth",
  "aws_cognito_user_pools",
  "aws_iam",
  "aws_lambda",
  "aws_oidc",
  "aws_subscribe",
]
`;

exports[`Directive Definitions include correct directives in V1 directives 1`] = `
Array [
  "auth",
  "connection",
  "function",
  "http",
  "key",
  "model",
  "predictions",
  "searchable",
  "versioned",
]
`;

exports[`Directive Definitions include correct directives in V2 directives 1`] = `
Array [
  "auth",
  "belongsTo",
  "default",
  "deprecated",
  "function",
  "hasMany",
  "hasOne",
  "http",
  "index",
  "manyToMany",
  "mapsTo",
  "model",
  "predictions",
  "primaryKey",
  "refersTo",
  "searchable",
  "sql",
]
`;

exports[`Directive Definitions include correct directives in default directives 1`] = `
Array [
  "aws_api_key",
  "aws_auth",
  "aws_cognito_user_pools",
  "aws_iam",
  "aws_lambda",
  "aws_oidc",
  "aws_subscribe",
  "auth",
  "belongsTo",
  "default",
  "deprecated",
  "function",
  "hasMany",
  "hasOne",
  "http",
  "index",
  "manyToMany",
  "mapsTo",
  "model",
  "predictions",
  "primaryKey",
  "refersTo",
  "searchable",
  "sql",
]
`;
